### PR TITLE
[subset] set emptied CFF charstrings to 'endchar' with --retain-gids

### PR DESCRIPTION
**Pull request by anthrotype**
Part of #1446

is it just setting the program to `["endchar"]` or am I missing something?

### CODE REVIEW

File: Lib/fontTools/subset/cff.py [hunk 1]
```diff
@@ -104,37 +104,44 @@ def subset_glyphs(self, s):
...
+					c = cs[g]
+					c.decompile()
+					c.program = ["endchar"]
 		else:
-			cs.charStrings = {g:v
-					  for g,v in cs.charStrings.items()
-					  if g in s.glyphs}
-		font.charset = [g for g in font.charset if g in s.glyphs]
-		font.numGlyphs = len(font.charset)
+			# Load all glyphs
```
Comment (anthrotype) [inline Lib/fontTools/subset/cff.py]: i didn't touch this whole block, simply indented

**Comment by miguelsousa** [comment]
I think the modified glyphs then inherit the defaultWidth as their advance width, so there may be more work needed.

**Comment by anthrotype** [comment]
oh right, so we need to do something similar to what we already do when we empty the .notdef glyph.
i'll do that after lunch, thanks

**Comment by anthrotype** [comment]
On the other hand, given that these glyphs will be unreachable from cmap, hence virtually inexistent, I wonder if that matters...

**Comment by miguelsousa** [comment]
Yeah Cosimo, you can't think about encoded glyphs only 😉

**Comment by khaledhosny** [comment]
I *think* the glyph ids are simply kept so that the subset font have the same glyph ids as the original ones, all other properties are of not much interest.

**Comment by miguelsousa** [comment]
But the glyphs are still in the font, so to make them appear non-existent (i.e "subset") most everything about them should be zeroed, I think.

**Comment by readroberts** [comment]
@anthrotype Don't forget that CFF2 charstrings have no width and no 'endchar'.

**Comment by anthrotype** [comment]
Ok but #1448

**Comment by anthrotype** [comment]
@readroberts ok, done with https://github.com/anthrotype/fonttools/commit/9b296463bde96e6032dedd4c886580cae8ee171e -- which also fixed the issue in #1448. 
I think this PR is ready now

File: Lib/fontTools/subset/cff.py [hunk 2]
```diff
@@ -104,37 +116,41 @@ def subset_glyphs(self, s):
...
-				# https://github.com/khaledhosny/ots/pull/31
-				#sel.format = None
-				sel.format = 3
-				sel.gidArray = [sel.gidArray[i] for i in indices]
-			cs.charStrings = {g:indices.index(v)
-					  for g,v in cs.charStrings.items()
-					  if g in s.glyphs}
+		if s.options.retain_gids:
+			for g in s.glyphs_emptied:
+				_empty_charstring(font, g, isCFF2=cff.major > 1, width=0)
```
Comment (behdad) [inline Lib/fontTools/subset/cff.py]: I think we want to pass private.defaultWidthX here.

Comment (anthrotype) [reply Lib/fontTools/subset/cff.py]: I don't understand.. i'm passing the cff font in as parameter, then for CFF 1.0 I'm getting it's private.defaultWidthX and compute/store the charstring width. I thought for CFF2 we never want to store the width.

**Comment by behdad** [comment]
> I think the modified glyphs then inherit the defaultWidth as their advance width, so there may be more work needed.

That's fine to me, and in fact, we should do exactly that.

**Comment by anthrotype** [comment]
> That's fine to me, and in fact, we should do exactly that.

I had a feeling you preferred that way... Want me to revert/simplify?

File: Lib/fontTools/subset/cff.py (hunk 2 above)
Comment (behdad) [reply Lib/fontTools/subset/cff.py]: You're passing width=0, which requires storage in the charstring in CFF 1.0.  We don't care about width of the emptied glyphs, so we want to not store anything, which is equivalent to default width.  For CFF2 of course, there is no width whatsoever.

**Comment by behdad** [comment]
> > That's fine to me, and in fact, we should do exactly that.
> 
> I had a feeling you preferred that way... Want me to revert/simplify?

Yes please.

**Comment by anthrotype** [comment]
yeah, I did that intentionally because others pointed that out, but I realise it makes more sense to not waste storage for data which is not going to be used

**Comment by behdad** [comment]
> yeah, I did that intentionally because others pointed that out, but I realise it makes more sense to not waste storage for data which is not going to be used

That, and OpenType 1.6 on say width comes from hmtx, not CFF table.

**Comment by anthrotype** [comment]
it would be tempting to also drop the custom width of the ".notdef" glyph, but I kept it as it was (it's just one glyph after all).

**Pull request merged**